### PR TITLE
AI-618 , AI-665 : attachment for new forms

### PR DESCRIPTION
**Pull request by yuriyz**

### CODE REVIEW

File: core/model/src/main/java/org/activityinfo/model/resource/Resources.java [hunk 1]
```diff
@@ -60,6 +60,13 @@ public static Resource fromJson(String json) {
         return fromJson(resourceObject);
     }
 
+    public static Record fromJson_(String json) {
```
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/resource/Resources.java]: Improve naming here. For example, rename to recordFromJson() and resourceFromJson()

File: core/model/src/main/java/org/activityinfo/model/type/attachment/AttachmentType.java [hunk 2]
```diff
@@ -68,22 +71,38 @@ public FormClass getParameterFormClass() {
...
 
+            EnumType kindType = (EnumType) EnumType.TYPE_CLASS.createType();
+            kindType.getValues().add(new EnumItem(ResourceId.valueOf(Kind.ATTACHMENT.name()), "Attachment"));
+            kindType.getValues().add(new EnumItem(ResourceId.valueOf(Kind.IMAGE.name()), "Image"));
+
             FormClass formClass = new FormClass(ResourceIdPrefixType.TYPE.id("image"));
             formClass.addElement(new FormField(ResourceId.valueOf("cardinality"))
                     .setType(cardinalityType)
                     .setLabel("Cardinality")
                     .setDescription("Determines whether users can add a single image, or multiple images")
```
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/type/attachment/AttachmentType.java]: All user-readable strings should be internationalized.

File: core/model/src/main/java/org/activityinfo/model/type/attachment/AttachmentType.java [hunk 3]
```diff
@@ -68,22 +71,38 @@ public FormClass getParameterFormClass() {
...
+                            .setDescription("Determines type of attachment (image or file)")
+                            .setVisible(false)
+            );
             return formClass;
         }
     }
 
     public static final TypeClass TYPE_CLASS = new TypeClass();
 
+    public enum Kind {
```
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/type/attachment/AttachmentType.java]: Why not just parametrize on mime type? Images could have the mimetype `image/*`, otherwise any attachment would be "*/*".

File: core/model/src/main/java/org/activityinfo/model/util/Holder.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ * You should have received a copy of the GNU General Public 
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
+/**
+ * @author yuriyz on 11/17/2015.
+ */
+public class Holder<T> {
```
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/util/Holder.java]: How is this different than the `Optional` class?

File: server/src/main/java/org/activityinfo/service/blob/GcsBlobFieldStorageService.java [hunk 5]
```diff
@@ -84,57 +88,53 @@ public void put(AuthenticatedUser authenticatedUser, String contentDisposition,
...
-                             @PathParam("resourceId") ResourceId resourceId,
-                             @PathParam("fieldId") ResourceId fieldId,
                              @PathParam("blobId") BlobId blobId) throws IOException {
-        /* TODO: Ensure that users can download images they've just uploaded
-        ImageRowValue imageRowValue = getImageRowValue(user, resourceId, fieldId, blobId);
-        */
         GcsFilename gcsFilename = new GcsFilename(bucketName, blobId.asString());
         GcsService gcsService = GcsServiceFactory.createGcsService(RetryParams.getDefaultInstance());
         GcsInputChannel gcsInputChannel = gcsService.openPrefetchingReadChannel(gcsFilename, 0, ONE_MEGABYTE);
+        GcsFileMetadata metadata = gcsService.getMetadata(gcsFilename);
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/service/blob/GcsBlobFieldStorageService.java]: How are we verifying that the user has the right to access this blob?

File: server/src/main/java/org/activityinfo/service/blob/GcsBlobFieldStorageService.java [hunk 6]
```diff
@@ -43,23 +40,31 @@
...
-        LOGGER.info("Service account: " + appIdentityService.getServiceAccountName());
+        try {
+            LOGGER.info("Service account: " + appIdentityService.getServiceAccountName());
+        } catch (ApiProxy.CallNotFoundException e) {
+            // ignore: fails in local tests, bug in LocalServiceTestHelper?
+        }
     }
 
+    @GET
+    @Path("{blobId}/blob_url")
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/service/blob/GcsBlobFieldStorageService.java]: I believe all of our REST urls use camel case, while this end point uses underscores. Can we consistenly use camelCase?

File: server/src/main/java/org/activityinfo/service/blob/GcsBlobFieldStorageService.java [hunk 7]
```diff
@@ -84,57 +88,53 @@ public void put(AuthenticatedUser authenticatedUser, String contentDisposition,
...
         }
     }
 
     @GET
-    @Path("{resourceId}/{fieldId}/{blobId}/thumbnail")
+    @Path("{blobId}/image_url")
+    @Override
+    public Response getImageUrl(@InjectParam AuthenticatedUser user,
+                                @PathParam("blobId") BlobId blobId) throws IOException {
+        ImagesService imagesService = ImagesServiceFactory.getImagesService();
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/service/blob/GcsBlobFieldStorageService.java]: How are we verifying that the user has the right to access this blob?

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/AttachmentUploadRow.java [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        }
+
+        if (value.get().getMimeType().contains("pdf")) {
+            appendTumbnailImage("icons.filePdf");
+        } else if (value.get().getMimeType().startsWith("text")) {
+            appendTumbnailImage("icons.file");
+        }
+    }
+
+    private void appendTumbnailImage(String iconClassName) {
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/AttachmentUploadRow.java]: s/Tumbnail/Thumbnail

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+                // event.getResults is always null because of cross-domain upload
+                // we are forced to make additional call to check whether upload is successful
+
+                fetchImageServingUrl();
+            }
+        });
+        formPanel.submit();
+    }
+
+    public void fetchImageServingUrl() {
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java]: Is it neccessary to fetch the serving url, or can we just set the  src value to the end point, and have the end point respond with 303 See other?

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            String fileName = fileName();
+            String mimeType = MimeTypeUtil.mimeTypeFromFileName(fileName);
+
+            attachment.get().setMimeType(mimeType);
+            attachment.get().setFilename(fileName);
+            attachment.get().setBlobId(blobId);
+            return "/service/blob/credentials/" + blobId;
+        } catch (Exception e) {
+            Log.error(e.getMessage(), e);
+            Window.alert(I18N.CONSTANTS.failedToCreateUploadUrl());
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java]: I'm not sure this error message is going to be useful for users. Can we just report that the upload failed? You can add more details in the logging.

Is there a retry path?

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        this.formPanel = formPanel;
+        this.fileUpload = fileUpload;
+        this.attachment = attachment;
+        this.hiddenFieldsContainer = hiddenFieldsContainer;
+        this.uploadCallback = uploadCallback;
+    }
+
+    public void requestUploadUrl() {
+        try {
+            RequestBuilder requestBuilder = new RequestBuilder(RequestBuilder.POST, URL.encode(createUploadUrl()));
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java]: Shouldn't the URL we receive from the server be encoded already?

File: server/src/main/java/org/activityinfo/ui/client/util/GwtUtil.java [hunk 12]
```diff
@@ -135,4 +139,15 @@ public static Rectangle getBsContainerRectangle(com.google.gwt.user.client.Eleme
         }
         return null;
     }
+
+    public static void removeChildren(ComplexPanel panel) {
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/util/GwtUtil.java]: Why is this needed rather than [ComplexPanel.clear()](http://www.gwtproject.org/javadoc/latest/com/google/gwt/user/client/ui/Panel.html#clear()) ?

File: server/src/test/java/org/activityinfo/service/blob/GcsBlobFieldStorageServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public 
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
+import com.google.appengine.repackaged.com.google.api.client.util.Strings;
```
Comment (akbertram) [inline server/src/test/java/org/activityinfo/service/blob/GcsBlobFieldStorageServiceTest.java]: Wrong import.

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java (hunk 10 above)
Comment (yuriyz) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java]: Yes, we can just report that upload failed. Background is that user may upload file which is not supported in by our MimeTypeUtil and therefore mimtype indentification fails. Like .png -> image/png. (it is very rare case because we have mapping for more than 200 file extension.) What retry path would you propose in case we fail to identify mime type ?

Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java]: Why should that lead to an error? We're just sniffing mime type based on extension, right? You can always return `application/octet-stream`.

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java (hunk 11 above)
Comment (yuriyz) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java]: It's url for requesting upload url (which is constructed manually in the same class).

File: core/model/src/main/java/org/activityinfo/model/util/Holder.java (hunk 4 above)
Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/util/Holder.java]: Optional does not allow you to set(), Holder class allows you to modify reference.

File: server/src/main/java/org/activityinfo/service/blob/GcsBlobFieldStorageService.java (hunk 5 above)
Comment (yuriyz) [inline server/src/main/java/org/activityinfo/service/blob/GcsBlobFieldStorageService.java]: whoops, there is no verification. Any user can access any file.

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java (hunk 10 above)
Comment (yuriyz) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java]: Agreed.

File: server/src/main/java/org/activityinfo/ui/client/util/GwtUtil.java (hunk 12 above)
Comment (yuriyz) [inline server/src/main/java/org/activityinfo/ui/client/util/GwtUtil.java]: Thanks! Removed redundant code in https://github.com/bedatadriven/activityinfo/commit/4e23656ef4f1b3da8c3dd7d51d55c7fe87a915ef

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java (hunk 9 above)
Comment (yuriyz) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java]: I like idea of fetching more because then we can log problems and also give feedback to user (Upload failed message).

**Comment by akbertram** [comment]
How are you storing the JSON for the field in MySQL?

**Comment by akbertram** [comment]
Do we have any server-side verification of image/attachment fields, or can I just push anything through?

**Comment by yuriyz** [comment]
We store json of attachment record in indicatorvalue.TextValue .

**Comment by yuriyz** [comment]
No, we don't have any server side verification whether it is valid attachment json as well as we don't check whether it contains valid values inside the record.

**Comment by akbertram** [comment]
@bddbot please build

**Comment by akbertram** [comment]
please build

File: core/model/src/main/java/org/activityinfo/model/util/Holder.java (hunk 4 above)
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/util/Holder.java]: I think that's one of the advantages of the `Optional` class. Is it possible to refactor your code to replace Holder<T> with Optional<T> ?

File: server/src/main/java/org/activityinfo/core/shared/util/MimeTypeUtil.java [hunk 14]
```diff
@@ -48,198 +49,224 @@
...
-//        xlm	application/vnd.ms-excel
-//        xls	application/vnd.ms-excel
-//        xlt	application/vnd.ms-excel
-//        xlw	application/vnd.ms-excel
-//        xof	x-world/x-vrml
-//        xpm	image/x-xpixmap
-//        xwd	image/x-xwindowdump
-//        z	application/x-compress
-//        zip	application/zip
+        fileExtensionToMimeTypeRegistry.put("txt", "text/plain");
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/core/shared/util/MimeTypeUtil.java]: Do we need to send all of this to client? Is it possible to have this done on the server, or to have GCS detect it on upload?

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+    @Override
+    public void setType(FieldType type) {
+        // ignore
+    }
+
+    @Override
+    public void clearValue() {
+        image.setUrl("");
+        servingUrl = null;
+        attachment.set(new Attachment());
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java]: Why not just `attachment = Optional.absent()` or `Optional.of(new Attachment())` ?

File: server/src/main/java/org/activityinfo/ui/client/page/entry/sitehistory/SiteHistoryRenderer.java [hunk 16]
```diff
@@ -62,6 +62,11 @@ public String render(ActivityFormDTO activityForm, List<LocationDTO> locations,
 
         boolean first = true;
         for (SiteHistoryDTO history : histories) {
+
+            if (isAttachmentHistoryEntry(history.getEntityName())) {
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/page/entry/sitehistory/SiteHistoryRenderer.java]: Isn't this skipping _all_ fields in an update if the update contains an attachment? If for example, I update a site, changing a quantity field _and_ an attachment, won't that skip both changes here? It seems like you should be just checking the indicator type of the field.

**Comment by akbertram** [comment]
Looking good, but still seems like we're missing quite a bit on authorization. Consider this: if I have edit permissions on a form, once I've saved an instance with a reference to an image blob, that image is now part of the submission and has to be accessible to everyone that would otherwise have access to form instance - the same that those would have access to the contents of a quantity or text field.

Here's an idea of how you could do this:
  1. When the user submits the form instance with the blob id reference, you need to verify (on the server) that the user submitting the instance has permission to read the blob. Otherwise anyone could take a given blob id and submit it with their own form to gain access to the blob.
  2. After the uploading the blob, the user should have access to the image for the purposes of previewing the image.
  3. Upon updating the instance with a new blob id, ownership of the blob needs to be transferred to the form.
  4. Subsequently, end point needs to verify that any user accessing the blob has read access to the corresponding form instance. (use `PermissionOracle`)
  3.

File: core/model/src/main/java/org/activityinfo/model/util/Holder.java (hunk 4 above)
Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/util/Holder.java]: It seems you don't like `Holder` class :). `Holder` class keeps reference on `Attachment` object, so we can get/set value from `FormFieldWidget` and pass it to `Uploader` (to set filename, mimetype etc). Advantages of immutable classes are clear but here idea of `Holder` is to have it mutable to share the reference.

https://github.com/bedatadriven/activityinfo/blob/ai-618/server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java#L290

Does it look tricky? We can replace `Holder` with `Optional` if add something like

```
class Attachment {
    void set(Attachment attachment) {
          // set all props of attachments
    }
}
```

Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/util/Holder.java]: Workaround is to keep attachment value in `Uploader` to avoid this cross-referencing but it does not look very natural to me.

Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/util/Holder.java]: Here is result without `Holder` keeping reference in `Uploader`: https://github.com/bedatadriven/activityinfo/commit/6f6afcbf2b6e012b6b61e4cb7f2570895ef163b4

File: server/src/main/java/org/activityinfo/ui/client/page/entry/sitehistory/SiteHistoryRenderer.java (hunk 16 above)
Comment (yuriyz) [inline server/src/main/java/org/activityinfo/ui/client/page/entry/sitehistory/SiteHistoryRenderer.java]: ah, yes. I will correct it.

File: server/src/main/java/org/activityinfo/core/shared/util/MimeTypeUtil.java (hunk 14 above)
Comment (yuriyz) [inline server/src/main/java/org/activityinfo/core/shared/util/MimeTypeUtil.java]: Not sure how to easily workaround it. We identify mimeType on client and put in `FieldValue`. (Which later goes to `FormInstance` if save is clicked). `MimeType` is not used at all during upload (it's used only on download). When we are intermediate state when upload is made but `FormInstance` is not saved then that mime-type is only on client.

File: server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/AttachmentUploadRow.java [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+        });
+        this.uploader.setAttachment(value);
+
+        fileUpload.addChangeHandler(new ChangeHandler() {
+            @Override
+            public void onChange(ChangeEvent event) {
+                if (fieldWidgetMode == FieldWidgetMode.NORMAL) {
+                    uploader.requestUploadUrl();
+                } else {
+                    Window.alert(I18N.CONSTANTS.uploadIsNotAllowedInDuringDesing());
```
Comment (akbertram) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/AttachmentUploadRow.java]: Can we just disable the controls in design mode rather than adding popups?

Comment (yuriyz) [inline server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/AttachmentUploadRow.java]: Agreed, I will remove alert.

**Comment by yuriyz** [comment]
@akbertram security topic seems to be addressed:
1) during `FormInstance` create/update checks that user has access to blob
2) blob keeps both: owner - form, creator - user
3) all blob access checks whether user has access to the form.

Are there anything else you have in mind?

**Comment by akbertram** [comment]
Sounds good. I think we need an acceptance test for security that executes against the API. Keeping attachments secure is something we need to be very careful about.

**Comment by yuriyz** [comment]
Security for blobs is checked at https://github.com/bedatadriven/activityinfo/blob/c22b64e928b4639e41a3f2cd84c7a12345412564/server/src/test/java/org/activityinfo/service/blob/GcsBlobFieldStorageServiceTest.java
1) positive tests
2) negative tests
3) blob security attack - https://github.com/bedatadriven/activityinfo/blob/c22b64e928b4639e41a3f2cd84c7a12345412564/server/src/test/java/org/activityinfo/service/blob/GcsBlobFieldStorageServiceTest.java#L210

**Comment by akbertram** [comment]
@yuriyz promising, but these really need to be end-to-end. Can you move them to tests/acceptance and actually hit the end points?
